src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -0,0 +1,256 @@
+                record.get_num_extents(),
+                current_segment->segment->get_segment_id(),
+                allocated_to);
+              return _write(t, record).si_then([this]() mutable {
+                return roll_segment(false).safe_then([this]() mutable {
+                  return (write_iertr::future<seastar::stop_iteration>)
+                    seastar::make_ready_future<
+                      seastar::stop_iteration>(seastar::stop_iteration::yes);
+                });
+              });
+            }
+            add_extent_to_write(record, extent);
+            it = extents.erase(it);
+          }
+          record_size_t rsize = record.get_encoded_record_length();
+          allocated_to += rsize.mdlength + rsize.dlength;
+
+          logger().debug(
+            "SegmentedAllocator::Writer::write: writing {} extents to segment {} at {}",
+            record.get_num_extents(),
+            current_segment->segment->get_segment_id(),
+            allocated_to);
+          return _write(t, record).si_then([] {
+            return (write_iertr::future<seastar::stop_iteration>)
+              seastar::make_ready_future<
+                seastar::stop_iteration>(seastar::stop_iteration::yes);
+          });
+        });
+      });
+    });
+  });
+}
+
+bool SegmentedAllocator::Writer::_needs_roll(segment_off_t length) const {
+  return allocated_to + length > current_segment->segment->get_write_capacity();
+}
+
+SegmentedAllocator::Writer::init_segment_ertr::future<>
+SegmentedAllocator::Writer::init_segment(Segment& segment) {
+  bufferptr bp(
+    ceph::buffer::create_page_aligned(
+      segment_manager.get_block_size()));
+  bp.zero();
+  auto header =segment_header_t{
+    journal.next_journal_segment_seq - 1, // current seg seq = next seg seq - 1
+    segment.get_segment_id(),
+    NO_DELTAS, 0, true};
+  logger().debug("SegmentedAllocator::Writer::init_segment: initting {}, {}",
+    segment.get_segment_id(),
+    header);
+  ceph::bufferlist bl;
+  encode(header, bl);
+  bl.cbegin().copy(bl.length(), bp.c_str());
+  bl.clear();
+  bl.append(bp);
+  allocated_to = segment_manager.get_block_size();
+  return segment.write(0, bl).handle_error(
+    crimson::ct_error::input_output_error::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error when initing segment"}
+  );
+}
+
+SegmentedAllocator::Writer::roll_segment_ertr::future<>
+SegmentedAllocator::Writer::roll_segment(bool new_rolling_fut) {
+  if (new_rolling_fut) {
+    rolling_segment = seastar::shared_promise<>();
+  }
+  assert(rolling_segment);
+  if (current_segment) {
+    auto fut = seastar::now();
+    if (!current_segment->inflight_writes.empty()) {
+      fut = seastar::when_all_succeed(
+        current_segment->inflight_writes.begin(),
+        current_segment->inflight_writes.end());
+    }
+    current_segment->outdated = true;
+    (void) fut.then(
```

Comment:
I think it is not a good idea leaving any future unlinked/ungated, e.g. shutdown won't be safe anymore.

Reply:
right, will correct it